core/templates/dev/head/services/StateTopAnswersStatsService.ts

```diff
@@ -43,15 +41,24 @@ oppia.factory('StateTopAnswersStatsService', [
      * @param {string} stateName
      */
     var refreshAddressedInfo = function(stateName) {
+      if (!stateTopAnswersStatsCache.hasOwnProperty(stateName)) {
+        return;
+      }
+
       var explorationId = ContextService.getExplorationId();
       var state = ExplorationStatesService.getState(stateName);
+      var stateStats = stateTopAnswersStatsCache[stateName];
+
+      if (stateStats.interactionId !== state.interaction.id) {
+        delete stateTopAnswersStatsCache[stateName];
+        return;
+      }
+
       var interactionRulesService = $injector.get(
```

Comment:
What if state.interaction.id is null? Should add test for this case.

Reply:
Seems like an error that state interaction should handle? This service only copies the ID from the state, so if there were already a state with an ID of null, that seems like a bigger problem?

Reply:
Interaction ID can be null, right? (At least for unpublished explorations -- e.g. when I first create an exploration, the initial card has no interaction yet.)

Reply:
Yeah, but we only copy interaction IDs from the backend, which come directly from the storage. This makes me think it'll always be non-null.

Reply:
No, what I'm saying is that it can legitimately be null. E.g. delete an interaction from any of the states -- the interaction ID is null after the deletion.

Reply:
Ah, so we store states with null interaction ids in the backend?

Reply:
You should probably double-check, but yes, I think so?

Reply:
I realized we should be able to use `InteractionAuditOneOffJob` to detect for null interaction ids, it groups them by it. Let's run it on production and see whether any states group under `None`?

Reply:
It can be null. Here is a way you can reproduce it:

1) Log in as an admin
2) Go to the Roles tab on /admin and make yourself an admin by updating the role.
3) Create an exploration with TextInput leading to an END state, and publish it.
4) Get some stats.
5) Unpublish the exploration.
6) Delete the first interaction and save changes.

You will see (e.g. by looking in the History tab) that the first state now has an interaction ID of null. This is a valid case that can occur in production.

Reply:
Gotcha, done and addressed.